### ISSUE TITLE
The results of 'where' clause in FLWOR are different in memory and database.

### ISSUE THREAD

**Issue by jesselet** [eXist-db/exist#2817]
### What is the problem
<pre><code>
for $i in $data1//item, $j in $data2//item
where $i/name = $j/name
return $i/name
</code></pre>
The results of 'where' clause in FLWOR are different in memory and database.
### What did you expect
The results of 'where' clause in FLWOR are the same in memory and database.
### Describe how to reproduce or add a test
<pre><code>
xquery version "3.0";
let $col := "/db/apps/test"
let $data1 :=
&lt;data>
    &lt;item>&lt;name>1&lt;/name>&lt;/item>
    &lt;item>&lt;name>2&lt;/name>&lt;/item>
&lt;/data>
let $data2 :=
&lt;data>
    &lt;item>&lt;name>2&lt;/name>&lt;/item>
    &lt;item>&lt;name>3&lt;/name>&lt;/item>
&lt;/data>
let $test := function ($data1, $data2) {
    for $i in $data1//item, $j in $data2//item
    where $i/name = $j/name
    return $i/name
}
return (
    try {
        $test($data1, $data2)
    } catch * { "test1: " || $err:description },
    (:in eXist 2.2, 3.6.1, 4.6.1, 4.7.0: OK:)

    try {
        xmldb:create-collection("", $col),
        let $data1 := doc(xmldb:store($col, "data1.xml", $data1))
        let $data2 := doc(xmldb:store($col, "data2.xml", $data2))
        return $test($data1, $data2)
    } catch * { "test2: " || $err:description }
    (:in eXist 2.2: OK:)
    (:in eXist 3.6.1, 4.6.1, 4.7.0: EXCEPTION: variable '$j' is not set:)
)
</code></pre>
### Context information
- OS: Windows 7
- Java: jre1.8.0_202
- eXist-db: 2.2, 3.6.1, 4.6.1, 4.7.0

**Comment by duncdrum** [comment]
yup confirmed, and looks like a bug that has been there for a while, ouch. 

tested on `5.0.0-RC8-SNAPSHOT` via docker

**Comment by line-o** [comment]
I would also see this as a bug.
It can be worked around with:
```
    let $j := $data2//item
    return
    for $i in $data1//item
    where $i/name = $j/name
    return $i/name
```

**Comment by line-o** [comment]
The equivalent form described in the [spec](https://www.w3.org/TR/xquery-31/#id-xquery-for-clause) does also fail:
```
    for $i in $data1//item
    for $j in $data2//item
    where $i/name = $j/name
    return $i/name

```
